src/compiler/lowering/SpecializeGenericFunctionsPass.cpp

```diff
@@ -179,40 +179,56 @@ namespace {
 
         /**
          * @brief Try to reuse an existing specialization for the given template function
-         * @param callOp The call operation
-         * @param templateFunction The template function called by the `callOp`
+         * @param operandTypes Operand types of the operation
+         * @param templateFunction The template function called by the operation
          * @return either an existing and matching `FuncOp`, `nullptr` otherwise
          */
-        FuncOp tryReuseExistingSpecialization(daphne::GenericCallOp callOp, FuncOp templateFunction) {
+        FuncOp tryReuseExistingSpecialization(TypeRange operandTypes, FuncOp templateFunction) {
             auto eqIt = specializedVersions.equal_range(templateFunction.sym_name().str());
             for(auto it = eqIt.first ; it != eqIt.second ; ++it) {
                 auto specializedFunc = it->second;
 
-                if(callTypesMatchFunctionTypes(specializedFunc.getType(), callOp->getOperandTypes())) {
+                if(callTypesMatchFunctionTypes(specializedFunc.getType(), operandTypes)) {
                     // reuse existing specialized function
                     return specializedFunc;
                 }
             }
             return nullptr;
         }
 
+        /**
+         * @brief Try to reuse an existing specializtion if one exists, else creates a new 
+         *  specialization
+         * @param operandTypes Operand types of the operation
+         * @param calledFunction The function called by the operation
+         * @return A `FuncOp`for the specialization
+         */
+        FuncOp createOrReuseSpecialization(TypeRange operandTypes, FuncOp calledFunction) {
+            // check for existing specialization that matches
+            FuncOp specializedFunc = tryReuseExistingSpecialization(operandTypes, calledFunction);
+            if(!specializedFunc) {
+                // Create specialized function
+                auto specializedTypes =
+                    getSpecializedFuncArgTypes(calledFunction.getType(), operandTypes);
+                specializedFunc = createSpecializedFunction(calledFunction, specializedTypes);
+            }
+            return specializedFunc;
+        }
+
+        /**
+         * @brief Recursively specializes all functions within a `FuncOp` based on calls to the functions
+         * @param function The `FuncOp` to scan for function specializations
+         */
         void specializeCallsInFunction(FuncOp function) {
```

Comment:
@pdamme: To resolve the question left open in our discussion: `specializeCallsInFunction` does not introduce any casts nor does any mechanism for calls to UDFs. Hence, since `y` is of type `si64`  the following would also not compile:

```
def testfunc(i: ui64)
{
    x = i + 2;
    return x;
}

y = 10;
z = testfunc(y);

```

Reply:
Thanks for the information. We will probably rethink a few characteristics of UDFs in DAPHNE anyway, maybe we will add automatic casts of function parameters if they have a fully specified type. That behavior would be consistent with built-in functions. For now, it's fine as it is.